lib/cli/src/commands/run/mod.rs

```diff
@@ -140,9 +143,81 @@ impl Run {
         #[cfg(not(feature = "sys"))]
         let engine = store.engine().clone();
 
-        let runtime =
-            self.wasi
-                .prepare_runtime(engine, &self.env, runtime, preferred_webc_version)?;
+        let registry_name = self
+            .env
+            .registry_public_url()?
+            .host_str()
+            .unwrap_or("unknown_registry")
+            .replace('.', "_");
+
+        // We don't have the bytes of the module yet, but we still want to have the
+        // package-capabilities cache be as close to an actual identifier as possible.
+        let package_cache_path = match &self.input {
+            PackageSource::File(f) => {
+                let full_path = f.canonicalize()?.to_path_buf();
+                let metadata = full_path
+                    .parent()
+                    .ok_or(anyhow!("No parent!"))?
+                    .metadata()?
+                    .modified()?;
+
+                let mut hash = twox_hash::XxHash64::default();
+                full_path.hash(&mut hash);
+                metadata.hash(&mut hash);
+
+                format!("path_{}.json", hash.finish())
+            }
+            PackageSource::Dir(f) => {
+                let full_path = f.canonicalize()?.to_path_buf();
+                let metadata = full_path.metadata()?.modified()?;
+
+                let mut hash = twox_hash::XxHash64::default();
+                full_path.hash(&mut hash);
+                metadata.hash(&mut hash);
+
+                format!("path_{}.json", hash.finish())
+            }
+            PackageSource::Package(p) => match p {
+                PackageSpecifier::Ident(id) => {
+                    format!("id_{id}.json")
+                }
+                PackageSpecifier::Path(f) => {
+                    let full_path = PathBuf::from(f).canonicalize()?.to_path_buf();
+                    let mut h = twox_hash::XxHash64::default();
+                    full_path.hash(&mut h);
+
+                    if full_path.is_dir() {
+                        full_path.metadata()?.modified()?.hash(&mut h);
+                    } else if full_path.is_file() {
+                        full_path
+                            .parent()
+                            .ok_or(anyhow!("No parent!"))?
+                            .metadata()?
+                            .modified()?
+                            .hash(&mut h);
+                    }
+                    format!("path_{}.json", h.finish())
+                }
+                PackageSpecifier::Url(u) => {
+                    let mut h = twox_hash::XxHash64::default();
+                    u.hash(&mut h);
+                    format!("path_{}.json", h.finish())
+                }
+            },
+        };
+
+        let runtime = self.wasi.prepare_runtime(
+            engine,
+            &self.env,
+            &self
+                .env
+                .cache_dir()
```

Comment:
nit: Can we move this path construction to a dedicated helper method somewhere?

Reply:
[5b8321d](https://github.com/wasmerio/wasmer/pull/4975/commits/5b8321d098b02353e0a960f805851a1456a29338)